### PR TITLE
Remove dependency on MIQEncode in MiqVimDatastore

### PR DESCRIPTION
**Pull request by agrare**
Remove dependency on `manageiq-gems-pending` for base64encode

### CODE REVIEW

**Comment by bdunne** [comment]
I'm :+1: for dropping the dependency, but...
```ruby
MIQEncode.base64Decode(Base64.encode64("abc"))
=> "YWJj\n"
Base64.decode64(MIQEncode.base64Encode("abc"))
=> "i\xB7"
```

File: lib/gems/pending/VMwareWebService/MiqVimDataStore.rb [hunk 1]
```diff
@@ -287,7 +286,7 @@ def get_file_content(filepath)
     /(\d*)\.(\d*)/ =~ @invObj.about['version']
     raise "get_file_content not supported on [#{@invObj.about['fullName']}]" if $1.to_i < 3 || ($1.to_i == 3 && $2.to_i < 5)
 
-    fileUrl = "https://#{@invObj.server}/folder/#{MIQEncode.base64Encode(filepath)}?dsName=#{MIQEncode.base64Encode(@name)}"
+    fileUrl = "https://#{@invObj.server}/folder/#{Base64.encode64(filepath)}?dsName=#{Base64.encode64(@name)}"
```
Comment (bdunne) [inline lib/gems/pending/VMwareWebService/MiqVimDataStore.rb]: I'm :+1: for dropping the dependency, but...
```ruby
MIQEncode.base64Decode(Base64.encode64("abc"))
=> "YWJj\n"
Base64.decode64(MIQEncode.base64Encode("abc"))
=> "i\xB7"
```

Apparently this is using URI.encode and decode.  Thanks for finding this, I probably would have hit it in the future.

I think you want
```ruby
URI.encode(data, /[^#{URI::PATTERN::UNRESERVED}]/)
```

Comment (agrare) [inline lib/gems/pending/VMwareWebService/MiqVimDataStore.rb]: Gah you're right, badly named method :(
What do you think about removing `base64Encode` from `MIQEncode` then?  I don't see any other callers of it [ref](https://github.com/search?utf8=%E2%9C%93&q=org%3AManageIQ+MIQEncode.base64Encode&type=Code) [ref](https://github.com/search?utf8=%E2%9C%93&q=org%3AManageIQ+MIQEncode.base64Decode&type=Code)

File: lib/gems/pending/VMwareWebService/MiqVimDataStore.rb [hunk 2]
```diff
@@ -287,7 +286,10 @@ def get_file_content(filepath)
     /(\d*)\.(\d*)/ =~ @invObj.about['version']
     raise "get_file_content not supported on [#{@invObj.about['fullName']}]" if $1.to_i < 3 || ($1.to_i == 3 && $2.to_i < 5)
 
-    fileUrl = "https://#{@invObj.server}/folder/#{MIQEncode.base64Encode(filepath)}?dsName=#{MIQEncode.base64Encode(@name)}"
+    file_path_url = URI.encode(filepath, /[^#{URI::PATTERN::UNRESERVED}]/)
+    ds_name_url = URI.encode(@name, /[^#{URI::PATTERN::UNRESERVED}]/)
+
+    fileUrl = "https://#{@invObj.server}/folder/#{file_path_url}?dsName=#{ds_name_url}"
```
Comment (chessbyte) [inline lib/gems/pending/VMwareWebService/MiqVimDataStore.rb]: Would doing something like the following be cleaner?
```ruby
fileUrl = Uri.new("https", nil, @invObj.server, ...).to_s
```

File: lib/gems/pending/VMwareWebService/MiqVimDataStore.rb [hunk 3]
```diff
@@ -287,7 +286,14 @@ def get_file_content(filepath)
...
 
-    fileUrl = "https://#{@invObj.server}/folder/#{MIQEncode.base64Encode(filepath)}?dsName=#{MIQEncode.base64Encode(@name)}"
+    uri_parts = {
+      :scheme => "https",
+      :host   => @invObj.server,
+      :path   => "/folder/#{filepath}",
+      :query  => "dsName=#{@name}"
+    }
+
+    fileUrl = URI::Generic.build(uri_parts).to_s
```
Comment (bdunne) [inline lib/gems/pending/VMwareWebService/MiqVimDataStore.rb]: Building the URI like this is much nicer, but I think we still need to `URI.encode(data, /[^#{URI::PATTERN::UNRESERVED}]/)`.
```ruby
irb(main):001:0> MIQEncode.base64Encode("a%b&c")
=> "a%25b%26c"
irb(main):002:0> URI::Generic.build(:scheme => "https", :host => "aaa", :path => "/zzz/", :query => "dsname=a%b&c").to_s
=> "https://aaa/zzz/?dsname=a%b&c"
```
Also, do you know what special characters can be in the `filepath`?
```ruby
irb(main):003:0> URI::Generic.build(:scheme => "https", :host => "aaa", :path => "/zzz/x%y^z", :query => "dsname=a%b&c").to_s
URI::InvalidComponentError: bad component(expected absolute path component): /zzz/x%y^z
	from /home/bdunne/.rubies/ruby-2.3.3/lib/ruby/2.3.0/uri/generic.rb:769:in `check_path'
	from /home/bdunne/.rubies/ruby-2.3.3/lib/ruby/2.3.0/uri/generic.rb:817:in `path='
	from /home/bdunne/.rubies/ruby-2.3.3/lib/ruby/2.3.0/uri/generic.rb:191:in `initialize'
	from /home/bdunne/.rubies/ruby-2.3.3/lib/ruby/2.3.0/uri/generic.rb:136:in `new'
	from /home/bdunne/.rubies/ruby-2.3.3/lib/ruby/2.3.0/uri/generic.rb:136:in `build'
	from (irb):4
```

Comment (agrare) [inline lib/gems/pending/VMwareWebService/MiqVimDataStore.rb]: > Also, do you know what special characters can be in the filepath?

It is a VM name so some of the special characters are already URI escaped, but you're right it is definitely safer to `URI.encode()` first

**Pull request merged**